src/hotspot/share/gc/serial/cardTableRS.cpp

```diff
@@ -72,274 +72,58 @@ void CardTableRS::maintain_old_to_young_invariant(Generation* old_gen, bool is_y
-  void do_generation(Generation* gen) {
-    // Skip the youngest generation.
-    if (SerialHeap::heap()->is_young_gen(gen)) {
-      return;
+void CardTableRS::verify() {
+  class CheckForUnmarkedObjects : public ObjectClosure {
+    DefNewGeneration* _young_gen;
+    CardTableRS* _card_table;
+
+   public:
+    CheckForUnmarkedObjects() {
+      SerialHeap* heap = SerialHeap::heap();
+      _young_gen = heap->young_gen();
+      _card_table = heap->rem_set();
     }
-    // Normally, we're interested in pointers to younger generations.
-    VerifyCTSpaceClosure blk(_ct, gen->reserved().start());
-    gen->space_iterate(&blk, true);
-  }
-};
 
-void CardTableRS::verify_space(Space* s, HeapWord* gen_boundary) {
-  // We don't need to do young-gen spaces.
-  if (s->end() <= gen_boundary) return;
-  MemRegion used = s->used_region();
-
-  CardValue* cur_entry = byte_for(used.start());
-  CardValue* limit = byte_after(used.last());
-  while (cur_entry < limit) {
-    if (*cur_entry == clean_card_val()) {
-      CardValue* first_dirty = cur_entry+1;
-      while (first_dirty < limit &&
-             *first_dirty == clean_card_val()) {
-        first_dirty++;
-      }
-      // If the first object is a regular object, and it has a
-      // young-to-old field, that would mark the previous card.
-      HeapWord* boundary = addr_for(cur_entry);
-      HeapWord* end = (first_dirty >= limit) ? used.end() : addr_for(first_dirty);
-      HeapWord* boundary_block = s->block_start(boundary);
-      HeapWord* begin = boundary;             // Until proven otherwise.
-      HeapWord* start_block = boundary_block; // Until proven otherwise.
-      if (boundary_block < boundary) {
-        if (s->block_is_obj(boundary_block) && s->obj_is_alive(boundary_block)) {
-          oop boundary_obj = cast_to_oop(boundary_block);
-          if (!boundary_obj->is_objArray() &&
-              !boundary_obj->is_typeArray()) {
-            guarantee(cur_entry > byte_for(used.start()),
-                      "else boundary would be boundary_block");
-            if (*byte_for(boundary_block) != clean_card_val()) {
-              begin = boundary_block + s->block_size(boundary_block);
-              start_block = begin;
-            }
-          }
-        }
-      }
-      // Now traverse objects until end.
-      if (begin < end) {
-        MemRegion mr(begin, end);
-        VerifyCleanCardClosure verify_blk(gen_boundary, begin, end);
-        for (HeapWord* cur = start_block; cur < end; cur += s->block_size(cur)) {
-          if (s->block_is_obj(cur) && s->obj_is_alive(cur)) {
-            cast_to_oop(cur)->oop_iterate(&verify_blk, mr);
-          }
-        }
+    void do_object(oop obj) override {
+      CheckForUnmarkedOops object_check(_young_gen, _card_table);
+      obj->oop_iterate(&object_check);
+      if (object_check.has_unmarked_oop()) {
+        guarantee(_card_table->is_dirty_for_addr(obj), "Found unmarked old-to-young pointer");
       }
```

Comment:
```suggestion
      guarantee(!object_check.has_unmarked_oop(), "Found unmarked old-to-young pointer");
```

This looks much more concise, there is no need to repeat the check imo. Please also add some information about the failing address into the message.

Reply:
The `if` is required to account for imprecise-mark -- obj has no unmarked oops or obj-start is dirty.